source/fundamentals/data-formats/documents.txt

```diff
@@ -14,34 +14,379 @@ Documents
+  MongoCollection<Document> collection = database.getCollection("authors");
+
+  InsertOneResult result = collection.insertOne(author);
+
+Once you perform a successful insert, you can retrieve the sample document
+data from the collection using the following code:
+
+.. code-block:: java
+
+   import com.mongodb.client.model.Filters;
+   // <MongoCollection setup code here>
+
+   Document doc = collection.find(Filters.eq("name", "Gabriel García Márquez")).first();
+   if (doc != null) {
+       System.out.println("_id: " + doc.getObjectId("_id")
+           + ", name: " + doc.getString("name")
+           + ", lastUpdated: " + doc.getDate("lastUpdated"));
+
+       doc.getList("novels", Document.class).forEach((novel) -> {
+           System.out.println("title: " + novel.getString("title")
+               + ", yearPublished: " + novel.getInteger("yearPublished"));
+           });
+   }
+
+.. note::
+
+   The code sample above uses helper methods that check the returned type
+   and throw an exception if it is unable to cast the field value.
+   You can call the :java-docs:`get() </apidocs/bson/org/bson/Document.html#get(java.lang.Object)>`
+   method specified by the ``Map`` interface to retrieve field values as
+   type ``Object`` and to skip type checking.
+
+The output should resemble the following:
+
+.. code-block:: none
+   :copyable: false
+
+   _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020
+   title: One Hundred Years of Solitude, yearPublished: 1967
+   title: Chronicle of a Death Foretold, yearPublished: 1981
+   title: Love in the Time of Cholera, yearPublished: 1985
+
+For more information on retrieving and manipulating MongoDB data, see our
+:doc:`CRUD guide </fundamentals/crud>`.
+
+For more information on this class, see the
+:java-docs:`Document API documentation </apidocs/bson/org/bson/Document.html>`.
 
 BsonDocument
 ------------
 
-- Description
-- Link to Fundamentals > Document Data Formats: BSON for more information
-- Code example of declaring one and appending fields
-- Link to Fundamentals > CRUD for more information
-- Link to API documentation
+The ``BsonDocument`` class provides a type-safe API to access and manipulate
+a BSON document. You need to specify the BSON type from the Java BSON
+library for each field. See the following table for mappings between
+frequently-used BSON and Java BSON library types:
+
+.. list-table::
+   :header-rows: 1
+   :stub-columns: 1
+
+   * - BSON type
+     - Java BSON library type
+   * - Array
+     - ``org.bson.BsonArray``
+   * - Binary
+     - ``org.bson.BsonBinary``
+   * - Boolean
+     - ``org.bson.Boolean``
+   * - Date (long value)
+     - ``org.bson.BsonDateTime``
+   * - Document
+     - ``org.bson.BsonDocument``
+   * - Double
+     - ``org.bson.BsonDouble``
+   * - Int32
+     - ``org.bson.BsonInt32``
+   * - Int64
+     - ``org.bson.BsonInt64``
+   * - Null
+     - ``null``
```

Comment:
`null` is actually not allowed.  Use the `org.bson.BsonNull` type

Reply:
Thanks, didn't see that type. Will update.